packages/connection-info/src/connection-secrets.ts

```diff
@@ -190,10 +181,62 @@ export function extractSecrets(connectionInfo: Readonly<ConnectionInfo>): {
   return { connectionInfo: connectionInfoWithoutSecrets, secrets };
 }
 
-function omitPropertiesWhoseValuesAreEmptyObjects<
-  T extends Record<string, Record<string, unknown>>
->(obj: T): Partial<T> {
+function omitPropertiesWhoseValuesAreEmptyObjects<T extends Document>(obj: T) {
   return Object.fromEntries(
     Object.entries(obj).filter(([, value]) => Object.keys(value).length > 0)
   ) as Partial<T>;
 }
+
+const KMS_PROVIDER_SECRET_PATHS = {
+  local: ['key'],
+  aws: ['secretAccessKey', 'sessionToken'],
+  azure: ['clientSecret'],
+  gcp: ['privateKey'],
+};
+
+type AutoEncryptionKMSAndTLSOptions = Partial<
+  Pick<AutoEncryptionOptions, 'kmsProviders' | 'tlsOptions'>
+>;
+
+function extractAutoEncryptionSecrets(data: AutoEncryptionOptions): {
+  data: AutoEncryptionOptions & AutoEncryptionKMSAndTLSOptions;
+  secrets: AutoEncryptionKMSAndTLSOptions;
+} {
+  const secrets: AutoEncryptionKMSAndTLSOptions = {};
+  // Secrets are stored in a kmsProviders and tlsOptions
+  const { kmsProviders, tlsOptions, ...result } = data;
+
+  for (const kmsProviderName in kmsProviders ?? {}) {
```

Comment:
```suggestion
  for (const kmsProviderName of Object.keys(kmsProviders ?? {})) {
```

Not a huge deal, but `for ... in` is generally considered a bit less reliable because it walks the prototype chain, so while that doesn't make a big difference here, it's something we shouldn't start making common use of in the codebase